### ISSUE TITLE
MCP-0006: Atomic Blocks

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1154]
**Modified by dietmarw on 2 Dec 2015 10:03 UTC**
Modelica Change Proposal Idea to support improved modularization of dynamic execution aspects of blocks. This is especially relevant in the context of automatic code generation for (safety relevant) embedded control applications (but also beneficial in more general settings).

### Documents
* https://svn.modelica.org/projects/MCP/MAinternal/MCP-0006_AtomicBlocks

----

**Reported by thiele on 2 Jun 2013 16:28 UTC**
Modelica Change Proposal Idea to support improved modularization of dynamic execution aspects of blocks. This is especially relevant in the context of automatic code generation for (safety relevant) embedded control applications (but also beneficial in more general settings).

Described in: MCPI-0006_AtomicBlocks

Stored under SVN in MCP Area
https://svn.modelica.org/projects/MCP-Ideas/MAinternal/

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1154

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 3 Dec 2014 08:22 UTC**
To me the idea is not clear regarding what it aims to achieve.

Atomic is often used on data-types to indicate that a variable is updated at once.
Here it is used to indicate that the block code is turned into one function, and executed without any other code interleaved.

The question is what happens if we have a multi-threaded execution of the model code; should the atomic block multi-threaded execution?

**Comment by modelica-trac-importer** [comment]
**Comment by thiele on 4 Dec 2014 09:22 UTC**
The aim is modular code generation in which modular structures
at the model level (blocks) would be mapped to modular structures at the target code level (C-function(s)). This is has advantages if automatically generated code shall replace handwritten code in an (embedded systems) software project. The reason why I thought it should rather be a language element than an annotation for code-generation is that "atomic" execution of blocks imposes stronger causality constraints, thus it has semantic consequences.

Here, atomic doesn't refer to multi-threaded execution aspects. I had single-threaded code in mind when writing the MCP. There is no need to stop multi-threaded execution while executing an atomic block.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Dec 2014 14:10 UTC**
Replying to [comment:2 thiele]:
> Here, atomic doesn't refer to multi-threaded execution aspects. I had single-threaded code in mind when writing the MCP. There is no need to stop multi-threaded execution while executing an atomic block.

In that case I believe both that the MCP should make that clearer, and also that we should use a different name, since "atomic" is often used for data-types in multi-threading environment where changes are "atomic" among the different threads.

If the block communicates externally using a sequence of non-pure calls it would be surprising that two "atomic" blocks could be run in parallel in a multi-threaded environment.

https://en.wikipedia.org/wiki/Linearizability states:
"In concurrent programming, an operation (or set of operations) is atomic, linearizable, indivisible or uninterruptible if it appears to the rest of the system to occur instantaneously. Atomicity is a guarantee of isolation from concurrent processes. Additionally, atomic operations commonly have a succeed-or-fail definition—they either successfully change the state of the system, or have no apparent effect."

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 2 Dec 2015 10:03 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 28 Feb 2017 12:30 UTC**
Decided to put on hold at design meeting, since they are not actively worked on.
Until we have a new stage "MCP-on-hold" they are put on Milestone "NeedsInput" (similarly as MCP-0025).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 28 Feb 2017 12:38 UTC**
To make it easy to get overview these tickets that are effectively on hold are closed as expired. They can be re-opened later if/when we start working on them.
